v0.48.0 crash for tfe_team_project_access and tfe_team_access

<!--
Hi there,

Thank you for opening an issue! Please note that we try to keep the this issue tracker reserved for
bug reports and feature requests related to the Terraform Cloud/Enterprise provider. If you know
your issue relates to the Terraform Cloud/Enterprise platform itself, please contact
tf-cloud@hashicorp.support. For general usage questions, please post to our community forum:
https://discuss.hashicorp.com.
-->

#### Terraform Cloud/Enterprise version

<!---
Are you using Terraform Cloud or Terraform Enterprise? If Terraform Enterprise, please include the version
Example: Terraform Enterprise v202104-1
-->
 Terraform Enterprise v202307-1

#### Terraform version
<!---
If you are using Terraform CLI, run `terraform version` to show the version.

If you are using Terraform Cloud or Terraform Enterprise, see the Terraform version being used by a
workspace in the workspace's Overview tab.

Paste the result between the ``` marks below.
-->

```plaintext
v1.5.4
```

#### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.
-->

```terraform
resource "tfe_team_access" "default" {
  access       = "write"
  team_id      = <team_id>
  workspace_id =  <workspace_id>
}

resource "tfe_team_project_access" "default" {
  project_id =  <project_id>
  team_id    = <team_id>
  access     = "admin"
}
```

#### Debug Output
<!--
Full debug output can be obtained by running Terraform with the environment variable `TF_LOG=trace`. Please create a GitHub Gist containing the debug output. Please do _not_ paste the debug output in the issue, since debug output is long.

Debug output may contain sensitive information. Please review it before posting publicly, and if you are concerned feel free to encrypt the files using the HashiCorp security public key.
-->

```plaintext
Stack trace from the terraform-provider-tfe_v0.48.0_x5 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xcc89e4]

goroutine 532 [running]:
github.com/hashicorp/terraform-provider-tfe/tfe.resourceTFETeamProjectAccessRead({0x1067760, 0xc0008d2de0}, 0xc000584d00, {0xe3d0c0?, 0xc00076e510?})
github.com/hashicorp/terraform-provider-tfe/tfe/resource_tfe_team_project_access.go:314 +0x624
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0xc0003f1ea0, {0x1067798, 0xc00039a3f0}, 0xd?, {0xe3d0c0, 0xc00076e510})
github.com/hashicorp/terraform-plugin-sdk/v2@v2.27.0/helper/schema/resource.go:750 +0x12e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc0003f1ea0, {0x1067798, 0xc00039a3f0}, 0xc000365c70, {0xe3d0c0, 0xc00076e510})
github.com/hashicorp/terraform-plugin-sdk/v2@v2.27.0/helper/schema/resource.go:1044 +0x59e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ReadResource(0xc00020ee88, {0x1067798?, 0xc00039a2d0?}, 0xc000251480)
github.com/hashicorp/terraform-plugin-sdk/v2@v2.27.0/helper/schema/grpc_provider.go:616 +0x4a5
github.com/hashicorp/terraform-plugin-mux/tf5muxserver.muxServer.ReadResource({0xc0003ddec0, 0xc00040c100, 0xc0003ddef0, 0xc0003ddf20, {0xc0003b2cc0, 0x3, 0x4}}, {0x1067798?, 0xc000997f50?}, 0xc000251480)
github.com/hashicorp/terraform-plugin-mux@v0.11.2/tf5muxserver/mux_server_ReadResource.go:29 +0x139
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ReadResource(0xc0003bd900, {0x1067798?, 0xc0009977a0?}, 0xc0008d2ba0)
github.com/hashicorp/terraform-plugin-go@v0.18.0/tfprotov5/tf5server/server.go:751 +0x4b1
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadResource_Handler({0xea0be0?, 0xc0003bd900}, {0x1067798, 0xc0009977a0}, 0xc000224c40, 0x0)
github.com/hashicorp/terraform-plugin-go@v0.18.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:386 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00028c1e0, {0x106cf20, 0xc0001401a0}, 0xc00072e120, 0xc000411f80, 0x1615970, 0x0)
google.golang.org/grpc@v1.56.1/server.go:1337 +0xdf0
google.golang.org/grpc.(*Server).handleStream(0xc00028c1e0, {0x106cf20, 0xc0001401a0}, 0xc00072e120, 0x0)
google.golang.org/grpc@v1.56.1/server.go:1714 +0xa2f
google.golang.org/grpc.(*Server).serveStreams.func1.1()
google.golang.org/grpc@v1.56.1/server.go:959 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
google.golang.org/grpc@v1.56.1/server.go:957 +0x18c

Error: The terraform-provider-tfe_v0.48.0_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

#### Expected Behavior
<!--
What should have happened?
-->
The code should work.
#### Actual Behavior
<!--
What actually happened?
-->
The code works with `v0.47.0` but not wirh `v0.48.0`
#### Additional Context
<!--
Are there anything atypical about your situation that we should know? For example: is Terraform running in a wrapper script or in a CI system? Are you passing any unusual command line options or environment variables to opt-in to non-default behavior?
-->
Although the impact resources are different, my issue looks very similar to https://github.com/hashicorp/terraform-provider-tfe/issues/995.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,8 @@
 ## Unreleased
 
+BUG FIXES:
+* `r/tfe_team_project_access`: Fixes a panic that occurs when the client is configured against an older TFE release, by @sebasslash [1011](https://github.com/hashicorp/terraform-provider-tfe/pull/1011)
+
 FEATURES:
 * `d/tfe_organization_membership`: Add `organization_membership_id` attribute, by @laurenolivia [997](https://github.com/hashicorp/terraform-provider-tfe/pull/997)
 * `d/tfe_variable_set`: Add `project_ids` attribute, by @Netra2104 [994](https://github.com/hashicorp/terraform-provider-tfe/pull/994)

--- a/docs/backwards-compatibility.md
+++ b/docs/backwards-compatibility.md
@@ -1,0 +1,68 @@
+# Maintaining Backwards Compatibility (Terraform Enterprise)
+
+## Terraform Enterprise Version Checking
+
+**Note:** If you are introducing a new resource that is only available at a certain TFE release, you do not need to perform any sort of checks.
+
+### Implied Check (Recommended Approach)
+
+The simplest solution when a particular attribute is not supported by a given TFE release is a `nil` check. These checks are TFE release agnostic and assume that if a field was not returned by the API, the TFE release does not support it. They are particularly important for fields that are structs, where attempting to dereference a nil pointer causes a panic.
+
+You can use this implied checks when:
+
+- The field is a pointer
+- When a `null` value is ignored by the API or by go-tfe (see if the struct tag has `omitempty`)
+
+**Example:**
+
+```go
+if tmAccess.ProjectAccess != nil {
+    projectAccess := []map[string]interface{}{{
+    		"settings": tmAccess.ProjectAccess.ProjectSettingsPermission,
+    		"teams":    tmAccess.ProjectAccess.ProjectTeamsPermission,
+    	}}
+    // Write project access to state
+    err := d.Set("project_access", projectAccess)
+}
+```
+
+### Explicit Enterprise Checks
+
+If a resource or attribute is **only** available in Terraform Enterprise, use the go-tfe helper [IsEnterprise()](https://pkg.go.dev/github.com/hashicorp/go-tfe#Client.IsEnterprise) to ensure the client is configured against a TFE instance. This check is derived from the `TFP-AppName` header that Terraform Cloud emits, of which if not present, indicates a Terraform Enterprise installation.
+
+```go
+config := meta.(ConfiguredClient)
+
+if config.Client.IsEnterprise() {
+    // do something with TFE only behavior
+}
+```
+
+### Documentation
+
+It is important to communicate with practitioners which resources and fields are supported for a particular TFE release.
+
+For a new resource, add the minimum release required to the top level documentation.
+
+**Example:**
+
+```md
+# my_new_resource
+
+Provides a my new resource.
+
+~> **NOTE:** Using this resource requires using the provider with Terraform Cloud or an instance of Terraform Enterprise at least as recent as v202302-1.
+```
+
+
+If an attribute has a TFE release constraint, add a second sentence to the attribute's description:
+
+```md
+## Argument Reference
+
+The following arguments are supported:
+
+* `foo` - (Required) Foo is bar.
+* `bar` - (Optional) Bar is foo.
+* `foobar` - (Optional) Foobar is barfoo. This attribute requires Terraform Cloud or an instance of Terraform Enterprise at least as recent as `v202302-1`.
+```

--- a/go.mod
+++ b/go.mod
@@ -11,8 +11,8 @@ require (
 	github.com/hashicorp/go-hclog v1.5.0 // indirect
 	github.com/hashicorp/go-multierror v1.1.1 // indirect
 	github.com/hashicorp/go-retryablehttp v0.7.4 // indirect
-	github.com/hashicorp/go-slug v0.12.0
-	github.com/hashicorp/go-tfe v1.32.0
+	github.com/hashicorp/go-slug v0.12.1
+	github.com/hashicorp/go-tfe v1.32.1
 	github.com/hashicorp/go-version v1.6.0
 	github.com/hashicorp/hcl v1.0.0
 	github.com/hashicorp/hcl/v2 v2.17.0 // indirect

--- a/mkdocs.yml
+++ b/mkdocs.yml
@@ -15,6 +15,7 @@ nav:
           - Testing: testing.md
           - Introducing Beta Features: beta.md
           - Debugging: debugging.md
+          - Backwards Compatibility: backwards-compatibility.md
       - Acceptance Tests: writing-acceptance-tests.md
       - Changelog Process: changelog-process.md
 

--- a/tfe/resource_tfe_team_project_access.go
+++ b/tfe/resource_tfe_team_project_access.go
@@ -310,29 +310,34 @@ func resourceTFETeamProjectAccessRead(ctx context.Context, d *schema.ResourceDat
 		d.Set("project_id", "")
 	}
 
-	project_access := []map[string]interface{}{{
-		"settings": tmAccess.ProjectAccess.ProjectSettingsPermission,
-		"teams":    tmAccess.ProjectAccess.ProjectTeamsPermission,
-	}}
+	// These two fields are only available in TFC and TFE v202308-1 and later
+	if tmAccess.ProjectAccess != nil {
+		project_access := []map[string]interface{}{{
+			"settings": tmAccess.ProjectAccess.ProjectSettingsPermission,
+			"teams":    tmAccess.ProjectAccess.ProjectTeamsPermission,
+		}}
 
-	workspace_access := []map[string]interface{}{{
-		"state_versions": tmAccess.WorkspaceAccess.WorkspaceStateVersionsPermission,
-		"sentinel_mocks": tmAccess.WorkspaceAccess.WorkspaceSentinelMocksPermission,
-		"runs":           tmAccess.WorkspaceAccess.WorkspaceRunsPermission,
-		"variables":      tmAccess.WorkspaceAccess.WorkspaceVariablesPermission,
-		"create":         tmAccess.WorkspaceAccess.WorkspaceCreatePermission,
-		"locking":        tmAccess.WorkspaceAccess.WorkspaceLockingPermission,
-		"move":           tmAccess.WorkspaceAccess.WorkspaceMovePermission,
-		"delete":         tmAccess.WorkspaceAccess.WorkspaceDeletePermission,
-		"run_tasks":      tmAccess.WorkspaceAccess.WorkspaceRunTasksPermission,
-	}}
-
-	if err := d.Set("project_access", project_access); err != nil {
-		return diag.Errorf("Error setting configuration of team project access %s: %v", d.Id(), err)
+		if err := d.Set("project_access", project_access); err != nil {
+			return diag.Errorf("Error setting configuration of team project access %s: %v", d.Id(), err)
+		}
 	}
 
-	if err := d.Set("workspace_access", workspace_access); err != nil {
-		return diag.Errorf("Error setting configuration of team workspace access %s: %v", d.Id(), err)
+	if tmAccess.WorkspaceAccess != nil {
+		workspace_access := []map[string]interface{}{{
+			"state_versions": tmAccess.WorkspaceAccess.WorkspaceStateVersionsPermission,
+			"sentinel_mocks": tmAccess.WorkspaceAccess.WorkspaceSentinelMocksPermission,
+			"runs":           tmAccess.WorkspaceAccess.WorkspaceRunsPermission,
+			"variables":      tmAccess.WorkspaceAccess.WorkspaceVariablesPermission,
+			"create":         tmAccess.WorkspaceAccess.WorkspaceCreatePermission,
+			"locking":        tmAccess.WorkspaceAccess.WorkspaceLockingPermission,
+			"move":           tmAccess.WorkspaceAccess.WorkspaceMovePermission,
+			"delete":         tmAccess.WorkspaceAccess.WorkspaceDeletePermission,
+			"run_tasks":      tmAccess.WorkspaceAccess.WorkspaceRunTasksPermission,
+		}}
+
+		if err := d.Set("workspace_access", workspace_access); err != nil {
+			return diag.Errorf("Error setting configuration of team workspace access %s: %v", d.Id(), err)
+		}
 	}
 
 	return nil

